src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -312,6 +312,104 @@ class TestRequestRateManager : public TestLoadManagerBase,
     StopWorkerThreads();
   }
 
+  /// Helper function that will setup and run a case to verify custom data
+  /// behavior
+  /// \param tensors Vector of input ModelTensors
+  /// \param json_str The custom data json text
+  /// \param expected_values Vector of expected input values for each inference
+  /// \param expect_init_failure True if InitManager is expected to throw an
+  /// error
+  /// \param expect_thread_failure True if the thread is expected to have
+  /// an error
+  void TestCustomData(
+      std::vector<ModelTensor>& tensors, const std::string json_str,
+      std::vector<std::vector<int32_t>>& expected_values,
+      bool expect_init_failure, bool expect_thread_failure)
+  {
+    params_.user_data = {json_str};
+
+    TestCustomDataSetup(tensors);
+
+    if (expect_init_failure) {
+      REQUIRE_THROWS_AS(
+          InitManager(
+              params_.string_length, params_.string_data, params_.zero_input,
+              params_.user_data, params_.start_sequence_id,
+              params_.sequence_id_range, params_.sequence_length,
+              params_.sequence_length_specified,
+              params_.sequence_length_variation),
+          PerfAnalyzerException);
+      return;
+    } else {
+      REQUIRE_NOTHROW(InitManager(
+          params_.string_length, params_.string_data, params_.zero_input,
+          params_.user_data, params_.start_sequence_id,
+          params_.sequence_id_range, params_.sequence_length,
+          params_.sequence_length_specified,
+          params_.sequence_length_variation));
+    }
+
+    auto thread_status = TestCustomDataRunThread();
+
+    if (!thread_status.IsOk() && !expect_thread_failure) {
+      std::cout << "Unexpected thread failure: " << thread_status.Message()
+                << std::endl;
+    }
+
+    REQUIRE(thread_status.IsOk() != expect_thread_failure);
+    if (!thread_status.IsOk()) {
+      return;
+    }
+
+    auto recorded_inputs{stats_->recorded_inputs};
+    std::vector<std::vector<int32_t>> recorded_values;
+
+    // Convert the recorded inputs into values, for both shared memory and non
+    // shared memory cases
+    //
+    if (params_.shared_memory_type != SharedMemoryType::NO_SHARED_MEMORY) {
```

Comment:
This all looks really complex to be in a test. When tests get too complex, how can we easily verify that the test is verifying things correctly?

Is there any way to move this logic checking somewhere closer to where the logic actually happens?

If that's not possible, can you make a ticket that briefly outlines future work for pushing this complex testing logic closer to wherever the relevant logic is and presumably where we don't currently have any testing?

Reply:
I've pushed this into a helper function